### PR TITLE
Ivy: misc fixes

### PR DESCRIPTION
**Pull request by IgorMinar**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts [hunk 1]
```diff
@@ -45,22 +48,46 @@ export class TypeScriptReflectionHost implements ReflectionHost {
...
-          typeValueExpr = typeNodeToValueExpr(node.type);
+      if (!hasInjectDecorator) {
+        // It's not possible to get a value expression if the parameter doesn't even have a type.
+        if (node.type !== undefined) {
+          // It's only valid to convert a type reference to a value reference if the type actually
+          // has a value declaration associated with it.
+          let type = this.checker.getTypeFromTypeNode(node.type);
+          let typeNode = node.type;
+
+          if (type.isUnion()) {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: I believe this was never supported before as such union types do not survive in the metadata that Typescript emits (for JIT purposes). @trotyl Could you chime in and refer to the Typescript issue, I can't seem to find it.

File: packages/core/src/render3/jit/injectable.ts [hunk 2]
```diff
@@ -27,6 +27,10 @@ export function compileInjectable(type: Type<any>, srcMeta?: Injectable): void {
   const meta: Injectable = srcMeta || {providedIn: null};
 
   let def: any = null;
+
+  // if NG_INJECTABLE_DEF is already defined on this class then don't overwrite it
+  if (type.hasOwnProperty(NG_INJECTABLE_DEF)) return;
```
Comment (JoostK) [inline packages/core/src/render3/jit/injectable.ts]: I guess this fix also has to be applied to all other decorators 👍 (in the case that decorators are present and the static field is already present on the symbol)

File: packages/core/src/render3/jit/module.ts [hunk 3]
```diff
@@ -24,6 +24,9 @@ const EMPTY_ARRAY: Type<any>[] = [];
  * This function automatically gets called when a class has a `@NgModule` decorator.
  */
 export function compileNgModule(moduleType: Type<any>, ngModule: NgModule): void {
+  if (!ngModule) {
```
Comment (JoostK) [inline packages/core/src/render3/jit/module.ts]: Set `{}` as default value for the `ngModule` param? This condition can't be true given the current type, a default value would properly indicate that module data may be missing.

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts [hunk 4]
```diff
@@ -45,22 +48,46 @@ export class TypeScriptReflectionHost implements ReflectionHost {
       const name = parameterName(node.name);
 
       const decorators = this.getDecoratorsOfDeclaration(node);
-
+      const hasInjectDecorator = !!(decorators && decorators.find(d => d.name === 'Inject'));
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: We should test for the decorator to originate from `@angular/core`. Looking at this more, seems like such analysis is inappropriate for the `ReflectionHost` abstraction and should instead be done by the callers (such as `getConstructorDependencies` which does have semantic knowledge of the Angular decorators).

[the remark this answers is not in the document]
Comment (IgorMinar) [reply packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: thanks

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts (hunk 1 above)
Comment (IgorMinar) [reply packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: we have two tests that depend on this behavior and while I'm not sure if we supported this in JIT we do currently support it in AOT so I'm afraid that we can't just break this behavior.

File: packages/core/src/render3/jit/injectable.ts (hunk 2 above)
Comment (IgorMinar) [reply packages/core/src/render3/jit/injectable.ts]: @JoostK can you send a PR for the other decorators?

File: packages/core/src/render3/jit/module.ts (hunk 3 above)
Comment (IgorMinar) [reply packages/core/src/render3/jit/module.ts]: done.

File: packages/core/src/render3/jit/injectable.ts (hunk 2 above)
Comment (JoostK) [reply packages/core/src/render3/jit/injectable.ts]: @IgorMinar Sure, sometime later this week.

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts (hunk 4 above)
Comment (IgorMinar) [reply packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: yeah. I agree. I don't like how the code is structured today. I think this code will need a bigger refactoring to layer things correctly.

**Comment by IgorMinar** [comment]
I'm closing this in favor of  #26871. I thought I could split up that PR and land part of it but even these tests depend on us correctly emitting factory/summary/flatindex shims which is not yet fixed, so they'll have to land together. :-(

File: packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts (hunk 1 above)
Comment (trotyl) [reply packages/compiler-cli/src/ngtsc/metadata/src/reflector.ts]: Seems to be late here, the issue could be https://github.com/Microsoft/TypeScript/issues/9916 or https://github.com/Microsoft/TypeScript/issues/12703 (or some other linked ones) 😄

**Pull request closed without merging**